http2: send GOAWAY properly & don't continue reading unnecessarily

apapirovski opened the pull request:
Fix an issue where http2 wasn't correctly submitting Goaway frames for the session, as well as handle the case where ECONNRESET occurs after such a Goaway frame is received.

Fixes: https://github.com/nodejs/node/issues/20705
Fixes: https://github.com/nodejs/node/issues/20750
Fixes: https://github.com/nodejs/node/issues/20850

<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

apapirovski commented:
Ok, I think there might be more going on here. Still fails although less frequently it seems. Will continue investigating.

apapirovski commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/14936/

apapirovski commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/14970/ and https://ci.nodejs.org/job/node-test-commit/18605/
Stress test CI: https://ci.nodejs.org/job/node-stress-single-test/1851/

apapirovski commented:
@nodejs/http2 @addaleax This PR is mostly getting there in fixing several important bugs in http2 but I'm tripped up on Windows and FreeBSD. Does anyone here have a Windows system handy? If so, could you build it with `--debug-http2` & run `NODE_DEBUG=http2 tools/test.py parallel/test-http2-client-upload-reject` (or whatever the Windows equivalent is) and post the debug output from any of the failing runs?

apapirovski commented:
Or if anyone knows how to run a build with debug output on CI, that would also be amazing.

Trott commented:
> Or if anyone knows how to run a build with debug output on CI, that would also be amazing.

@nodejs/build

Trott commented:
> > Or if anyone knows how to run a build with debug output on CI, that would also be amazing.
> 
> @nodejs/build 

@nodejs/testing too

mcollina commented:
I might be able to get the windows output next week, if it's still needed.

apapirovski commented:
Took me forever but managed to build Node on FreeBSD. Let's see what it says... :)

apapirovski commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/14979/
Stress test CI: https://ci.nodejs.org/job/node-stress-single-test/1853/

apapirovski commented:
Ok, we're mostly there. Windows is the only system outstanding and I actually think it's an issue with the two tests, rather than a bug in http2.

Long-term I'm planning to do a more significant refactoring of how http2 end/close/destroy flow works but for now this should be good enough.

apapirovski commented:
@nodejs/http2 @addaleax @BridgeAR this is now ready for review and resolves 3 separate bugs in http2.

apapirovski commented:
@nodejs/http2 I've fixed the bug from #20850 so please have another look. It's in a separate commit for ease of review. Since it also has to do with properly closing the stream, I figured it might as well go in this PR. I've modified an existing trailers test to sort of test for the bug but we can't replicate the exact conditions that lead to it in Safari, IE & Edge so it's not a 100% match.

CI: https://ci.nodejs.org/job/node-test-pull-request/14993/

apapirovski commented:
Also, anyone have experience in submitting bug reports to Safari & Edge? The way they handle TRAILERS frames is incorrect. I don't see anything in the spec re: empty TRAILER frames (only with END_STREAM flag set) not being correct. In fact, it specifically states:

> Header lists are collections of zero or more header fields.

So our behaviour was to spec but clearly certain browsers cannot handle it.

mcollina commented on test/parallel/test-http2-trailers.js:
```diff
@@ -18,6 +18,7 @@ server.on('stream', common.mustCall(onStream));
 function onStream(stream, headers, flags) {
   stream.on('trailers', common.mustCall((headers) => {
     assert.strictEqual(headers[trailerKey], trailerValue);
+    stream.end(body);
```
Is this needed? I do not think this is the API we should look for in this case. If we are piping a stream to (`origin.pipe(dest)` or `pipeline(origin, dest)`), that stream will end by default. Maybe we can wait with a `_final` handler?

apapirovski replied:
@mcollina If a user pipes a stream and have not tried to process incoming data then there's no way we can both close the stream properly and let them get the trailers.

apapirovski replied:
The alternative to how this test is modified is to call `resume` or bind a data handler.

mcollina replied:
This needs to be documented.

Trott replied:
> This needs to be documented.

@mcollina Do you mean a comment needs to be added to the test? Or something needs to be added to the `http2` documentation?

mcollina replied:
I think this should be documented in the documentation.

apapirovski replied:
I agree. Once that's added, do you approve for this to land @mcollina?

mcollina replied:
Yes, but I think trailers is something that we should go back to.

apapirovski replied:
@mcollina I think other than documenting we can't do much. The flow can go something like this:

1. The third party sends some data (perhaps it's even larger than overall `windowSize`)
2. We don't handle that data, so the third party never sends the rest of the data nor the trailers
3. We send a response and initiate closing the connection since we haven't shown any interest in consuming the incoming data.

With `http` that would look different because data is just written to the socket immediately and all of it can come through. There's no mechanism there to tell the other side not to write the full payload along with the trailers.

mcollina replied:
Ok

apapirovski commented:
@nodejs/http2 please have a look today / tomorrow if possible. This will immediately unbreak the CI and resolve a major bug in Safari, Edge & IE.

apapirovski commented:
Final CI: https://ci.nodejs.org/job/node-test-pull-request/15010/

apapirovski commented:
Landed in 4f0ab76b6c...ef9d0ae6f2
